Integrate solc test suite for hevm-symbolic

xwvvvvwx opened the pull request:
## Overview

This PR integrates two parts of the solc test suite and uses them to verify `hevm symbolic`:

1. the yul optimzer tests: a corpus of ~400 pairs of yul programs that have been written to be equivalant. Used to test `hevm equivalence`
2. the SMTCheker tests: a corpus of ~450 solidity contracts containing assertions, and annotations on whether or not those assertions should be violated. Used to test `hevm symbolic`

Not every test was useful in our context, but accounting for skipped tests, this gives us 474 new test cases.

The yul equivlance tests are run with both `cvc4` and `z3`, but the smt checker tests can only be run in  a reasonable time using `cvc4`. Interstingly `z3` is slightly faster on the equivalance tests (~30s for z3 and ~40s for cvc4 on my machine).

## Issues Found

During the course of integrating these tests I found 3 `hevm` bugs (including one soundness issue)

- https://github.com/dapphub/dapptools/issues/456
- https://github.com/dapphub/dapptools/issues/457
- https://github.com/dapphub/dapptools/issues/466

I also found four instances of incorrect tests in the yul optimizer test suite (https://github.com/ethereum/solidity/issues/9500).

The smtChecker tests are currently failing due to the soundness issue (https://github.com/dapphub/dapptools/issues/466).

## Results

*SMTChecker*:

```
ran: 451
passed: 288
skipped: 162
failed (smt reports assertion, hevm reports safe): 1
failed (hevm reports assertion, smt reports safe): 0
hevm timeout: 0 
hevm failure: 0
```

*yulEquivlence*

```
ran: 409
passed: 185
same bytecode: 89
ignored: 135
could not compile first program: 0
could not compile second program: 0
hevm timeout: 0
hevm execution failed: 0
```

cc @leonardoalt

leonardoalt commented on nix/hevm-tests/smt-checker.nix:
```diff
@@ -0,0 +1,383 @@
+    ${grep} -q 'Assertion checker does not yet support' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
+    ${grep} -q 'Internal error: Expression undefined for SMT solver.' $1
+    if [ $? == 0 ]; then ${echo} ${strings.smtCheckerFailed} && exit; fi
```
Do you need to exit if that's the case (for the block) above? It could still report some assertion violation

xwvvvvwx replied:
oh, I didn't realise that was the case. Removing this didn't change the numbers in the summary as the only new test executed was an SMTChecker false positive.

leonardoalt commented on overlay.nix:
```diff
@@ -21,9 +21,12 @@ in rec {
   buildDappPackage = import ./nix/build-dapp-package.nix { inherit (self) pkgs; };
 
   # Here we can make e.g. integration tests for Dappsys,
```
```suggestion
  # Here we can make e.g. integration tests for Dappsys.
```

MrChico commented on nix/hevm-tests/smt-checker.nix:
```diff
@@ -0,0 +1,383 @@
+    "inheritance/constructor_state_variable_init_chain_run_all.sol"
+    "inheritance/constructor_state_variable_init_chain_run_all_2.sol"
+
+    # --- unbounded looping ---
```
We can add `--max-iterations n` and try some of these. It looks like `functions/functions_recursive.sol` and `functions/functions_recursive_indirect` should work, even for `n = 1` for example.

MrChico commented:
This is super nice!
I have some reflections;
- First of all, I would expect ci to fail since the build is failing, but this doesn't seem to be the case. Do you know why?
- The summary seems particularly valuable, and it would be nice to have it easily accessible after a ci run. Maybe it would make sense to have these tests in their own ci workflow. 
Then we also wouldn't push them to the cache, because when we do have a cache hit, the test results are buried somewhere in a previous run. Or perhaps this is a good thing, because we don't have to wait for things that we know don't change. Regardless, it should be easy to find the summary.
- It could be nice to add some basic performance for the summary as well. Maybe we can just prefix all hevm invocations with `time` and add it up in some nice way in the end.

xwvvvvwx commented on nix/hevm-tests/smt-checker.nix:
```diff
@@ -0,0 +1,383 @@
+    "types/enum_explicit_values.sol"
+    "types/enum_transitivity.sol"
+
+    # --- smt checker false positives ---
```
What makes you say that?

leonardoalt replied:
the smtcheckertests has a bunch of other "assertion violation happens here" because of aliasing etc

xwvvvvwx replied:
Are those false positives? If I understand correctly, unless the function is marked as `external` aliasing issues could manifest from within the same contract, or from usage in an inherited contract, so I think I view those as valid warnings?

xwvvvvwx commented:
- fixed ci status (tests were not actuall being run in ci..)
- added `--max-iterations`
- added performance numbers to the summary

@MrChico regarding summary output / CI stages I'm not sure what the best approach is. Do you think it would be enough to run the hevm tests in the `run integration tests` section of the pipeline? Personally I think that not rerunning the tests if nothing has changed is quite nice, but maybe we can add some script that always prints the summary in CI even if the tests have not been re-executed?

MrChico commented:
I think we should ignore the soundness bug before merging this, so as to not block this PR.
This pipeline might warrant its own ci step

MrChico commented:
```
/nix/store/v5abzbplf5wflc4c47i5mg0s4hdznfcx-forceSymbolicMemory: line 20: /nix/store/wjpqpikc6szzv6lvgs984ilnwwv9vq4d-utillinux-1003.1-2008/bin/rev: No such file or directory
```
also these guys are in the logs

xwvvvvwx commented:
- ignored test failing due to soundness issue
- moved ci tests to their own stage
- used `rev` from `busybox` instead of `util-linux` (should be cross platform :crossed_fingers:)

leonardoalt commented:
Note for potentially another PR: all the tests in `test/libsolidity/semanticTests/*` that have the strings `// compileViaYul: also` or `// compileViaYul: true` could be compiled to Yul with (`solc --ir-optimized a.sol`) and without (`solc --ir`) the optimizer and equivalence checked.

xwvvvvwx commented:
@leonardoalt thx for the nice suggestion, I think that's out of scope for this PR atm. I've added it to the list at https://github.com/dapphub/dapptools/issues/468 to keep track of it.

xwvvvvwx commented:
CI is passing on both platforms now, I think this is good to go :slightly_smiling_face:

leonardoalt commented:
Amazing!
